### PR TITLE
cli/net_params: Warn on empty public-addr when starting a validator node

### PR DESCRIPTION
**Pull request by lexnv**
This PR shows a warning when the `--public-addr` is not provided for validators.

In the future, we'll transform this warning into a hard failure. Validators are encouraged to provide this parameter for better availability over the network.

cc @paritytech/networking

### CODE REVIEW

File: substrate/client/cli/src/config.rs [hunk 1]
```diff
@@ -185,7 +185,17 @@ pub trait CliConfiguration<DCV: DefaultConfigurationValues = ()>: Sized {
 			)
 		} else {
 			NetworkConfiguration::new(node_name, client_id, node_key, Some(net_config_dir))
-		})
+		};
+
+		if is_validator && network_config.public_addresses.is_empty() {
+			eprintln!(
```
Comment (dmitry-markin) [inline substrate/client/cli/src/config.rs:191]: It would be better to use `log::warn!` for this, to properly handle all the logging backends

**Review by dmitry-markin — APPROVED** [review]

File: substrate/client/cli/src/config.rs (hunk 1 above)
Comment (lexnv) [reply substrate/client/cli/src/config.rs:191]: Yep, although the logger is not initialized yet :(

We are fetching the config first, then initializing the logger with that config:

https://github.com/paritytech/polkadot-sdk/blob/dc5f96dc6166494c54feb92b0137265d657a0949/substrate/client/cli/src/lib.rs#L243-L245

Took some inspiration from:
https://github.com/paritytech/polkadot-sdk/blob/dc5f96dc6166494c54feb92b0137265d657a0949/substrate/client/cli/src/config.rs#L239-L245

Will look into initializing a temp logger, or remove the logging hook which uses this config cli 🤔

Comment (lexnv) [reply substrate/client/cli/src/config.rs]: Found an easier solution to this, print the warning after logging initialization.
After we do a release with this included, then we'll have to return in `network_config` an error and left a todo to not forget about it

**Review by dmitry-markin — APPROVED** [review]

File: substrate/client/cli/src/config.rs [hunk 2]
```diff
@@ -185,7 +185,12 @@ pub trait CliConfiguration<DCV: DefaultConfigurationValues = ()>: Sized {
 			)
 		} else {
 			NetworkConfiguration::new(node_name, client_id, node_key, Some(net_config_dir))
-		})
+		};
+
+		// TODO: Return error here in the next release.
```
Comment (skunert) [inline substrate/client/cli/src/config.rs:190]: We should have an issue for this and link the place where to add IMO. Already too many dead TODOs in the code :grimacing:

**Review by skunert — APPROVED** [review]

File: substrate/client/cli/src/config.rs [hunk 3]
```diff
@@ -638,6 +644,14 @@ pub trait CliConfiguration<DCV: DefaultConfigurationValues = ()>: Sized {
 
 		logger.init()?;
 
+		if config.role.is_authority() && config.network.public_addresses.is_empty() {
```
Comment (alexggh) [inline substrate/client/cli/src/config.rs:647]: Did you check parachain collators(is_authorithy returns true) won't get this warning?

Comment (xlc) [reply substrate/client/cli/src/config.rs:647]: https://github.com/paritytech/polkadot-sdk/issues/6280

**Pull request merged**